### ISSUE TITLE
noexcept TryGet/TrySet API consistency

### ISSUE THREAD

**Issue by alexey-romanov** [GiovanniDicanio/WinReg#53]
Hi.

I would like to suggest an API change for the exceptionless routines.

Currently if we examine the `TrySet*` and `TryGet*` routines it is seen that they have different semantics.
`TrySet*` return `RegResult` so it is possible to examine the error code.
`TryGet*` on other hand return `std::optinal<>`, so either result value in case of success or nothing in case of failure, making it impossible to examine the return code. 

I would suggest making so, that `TryGet*` would also report the `RegResult `code.

I would think of several ways to do so

1) Return `RegResult`, thus making `TryGet*` routines be consistent on return value with `TrySet*` routines

e.g.

```C++
[[nodiscard]] RegResult TryGetDwordValue(const std::wstring& valueName, DWORD * value) const;
```

2) Return `std::variant` holding either `RegResult `in case of failure or result value in case of success

e.g.

```C++
[[nodiscard]] std::variant<RegResult, DWORD> TryGetDwordValue(const std::wstring& valueName) const;
```

and with the following usage

```C++
RegKey key;
...

auto const res = key.TryGetDwordValue(L"BAMBAM");

if (std::holds_alternative<RegResult>(res))
{
    // failure here
    // print e.d. error code to stdout
    std::cout << std::get<RegResult>(res).code().value();
} else
{
    // Value query success
    const DWORD value = std::get<DWORD>(res);
}
```

3) Optionally provide a default-valued argument for ```RegResult``` storing if it is needed. This option is redundant from one side but form the other side is maybe more easy/convenient to use. 

```C++
[[nodiscard]] std::optional<DWORD> TryGetDwordValue(const std::wstring& valueName, RegResult * res = nullptr) const;
```

What do you think about the suggestion in general and about those options?

**Comment by GiovanniDicanio** [comment]
Thanks for your feedback. I know that returning `std::optional` looses the information about the error code. I believe this was already discussed in a past issue, and in that thread it was agreed to return `std::optional` to make the caller code simpler. In fact, compare the current `std::optional` pattern vs. your `std::holds_alternative<RegResult>/std::get<RegResult>/std::get<DWORD>` dance.

Maybe we could develop a tiny wrapper class that contains a `std::variant<DWORD, RegResult>`, and provides helper methods like `Suceeded()` that hide ugly calls to `std::holds_alternative<...>`, etc., and return instances of that class, to make the caller code easier, nicer and kind of less ugly.

Anyway, I'm currently busy with other stuff, and since [this project is not making me any money](https://blogs.msmvps.com/gdicanio/2022/03/30/open-source-who-pays-for-it/), I'm not much motivated to improving the current code right now.

**Comment by GiovanniDicanio** [comment]
Fixed by #56
